[AArch64] Allow peephole to optimize AND + signed compare with 0

AZero13 opened the pull request:
This should be the peephole's job. Because and sets V flag to 0, this is why signed comparisons with 0 are okay to replace with tst. Note this is only for AArch64, because ANDS on ARM leaves the V flag the same.

Fixes: https://github.com/llvm/llvm-project/issues/154387

AZero13 committed:
Pre-commit test (NFC)

AZero13 commented:
@davemgreen

AZero13 commented:
> I think the utility of your change would be more apparent if you pre-comitted some tests. IIUC this is folding things like:
> 
> ```
> 
>         and     w8, w9, w8
> 
>         cmp     w8, #0
> 
> ```
> 
> to
> 
> ```
> 
> tst w9, w8
> 
> ```
> 
> ?

Yes exactly that's the point

MacDue commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -1433,6 +1433,34 @@ static unsigned convertToNonFlagSettingOpc(const MachineInstr &MI) {
+    return AArch64::BICWrs;
+  case AArch64::BICSXrs:
+    return AArch64::BICXrs;
+  case AArch64::SBCSXr:
+    return AArch64::SBCXr;
+  case AArch64::SBCSWr:
+    return AArch64::SBCWr;
+  case AArch64::ADCSXr:
+    return AArch64::ADCXr;
```
Are all of these changes used? I don't see how SBC/ADC is relevant here?

MacDue commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -1921,7 +1998,8 @@ static bool canInstrSubstituteCmpInstr(MachineInstr &MI, MachineInstr &CmpInstr,
   // 2) If MI is add/sub with no-signed-wrap, it produces a poison value when
   //    signed overflow occurs, so CmpInstr could still be simplified away.
-  if (NZVCUsed->V && !MI.getFlag(MachineInstr::NoSWrap))
+  // 3) ANDS/BICS on AArch64 always sets V to 0, so it is safe to look through.
```
I think this case is already covered by `"MI and CmpInstr set N and V to the same value"`, which I think is true in this case.

AZero13 replied:
It's not.

SLE 0 is not MI

AZero13 replied:
AND instructions are not tagged as no overflow

And, the peephole as is checks for no touching of V and C.

However, and instructions on AArch64 (NOT ARM as in the 32 bit ISA) ALWAYS set C and V to 0.

As to when the C being 0 can be useful, I'm trying to figure that out and that's for another PR, but for now,know that cmp with 0 also always clears V, but the peephole doesn't know that.

AZero13 replied:
Ah I get what you mean, but I think it's important to clarify that AND does not set N and V to the same value, because AND sets V to 0 always, and N to if the value is negative.

AZero13 committed:
[AArch64] Allow peephole to optimize AND + signed compare with 0

This should be the peephole's job. Because and sets V flag to 0, this is why signed comparisons with 0 are okay to replace with tst. Note this is only for AArch64, because ANDS on ARM leaves the V flag the same.

Fixes: #154387

AZero13 commented:
@MacDue Done!

AZero13 commented:
@davemgreen

AZero13 commented:
@davemgreen @MacDue ping?

AZero13 commented:
@MacDue ping?

MacDue commented on llvm/lib/Target/AArch64/GISel/AArch64PostSelectOptimize.cpp:
```diff
@@ -87,6 +87,26 @@ unsigned getNonFlagSettingVariant(unsigned Opc) {
     return AArch64::ADDXri;
   case AArch64::ADDSWri:
     return AArch64::ADDWri;
+  case AArch64::ANDSWri:
```
This was marked as resolved and ignored, but this change does not relate to the rest of this PR (which is not in global ISEL). You can remove this, and all tests still pass. 

Your tests are only for a few AND cases, but you've implemented hooks for many more instructions that are untested. I think it would be better if you reduce the size of the PR to match the code paths you've tested.

AZero13 replied:
Many more? Just BICS.

MacDue reviewed (approved):
LGTM

AZero13 commented:
Thank you @MacDue can you please merge?
